src/sc.c

```diff
@@ -1,62 +1,291 @@
 #include "jemalloc/internal/jemalloc_preamble.h"
 
 #include "jemalloc/internal/assert.h"
+#include "jemalloc/internal/bit_util.h"
 #include "jemalloc/internal/sc.h"
-#include "jemalloc/internal/size_classes.h"
 
 sc_data_t sc_data_global;
 
+static size_t
+reg_size_compute(int lg_base, int lg_delta, int ndelta) {
+	return (ZU(1) << lg_base) + (ZU(ndelta) << lg_delta);
+}
+
+/* Returns the number of pages in the slab. */
+static int
+slab_size(int lg_page, int lg_base, int lg_delta, int ndelta) {
+	size_t page = (ZU(1) << lg_page);
+	size_t reg_size = reg_size_compute(lg_base, lg_delta, ndelta);
+
+	size_t try_slab_size = page;
+	size_t try_nregs = try_slab_size / reg_size;
+	size_t perfect_slab_size = 0;
+	bool perfect = false;
+	while (!perfect) {
```

Comment:
nit: add some comments about this being okay since in practice the common case is a few pages.

Reply:
Added a comment; the strict bound here is 2 * ngroup - 1.